src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2582,18 +2599,172 @@ CalculateNewConnectionCount(WorkerPool *workerPool)
+			 * connections required.
+			 */
+			if (ExecutorCostModel == EXECUTOR_COST_MODEL_CURRENT_EXECUTION_STATS &&
+				UsingExistingConnectionsCheaperThanEstablishingNewConnections(
+					newConnectionCount, workerPool))
+			{
+				return 0;
+			}
 		}
 	}
+
 	return newConnectionCount;
 }
 
 
+/*
+ * UsingExistingConnectionsCheaperThanEstablishingNewConnections returns true if
+ * using the already established connections takes less time compared to opening
+ * new connections based on the current execution's stats.
+ *
+ * The function returns false if the current execution has not established any connections
+ * or finished any tasks (e.g., no stats to act on).
+ */
+static bool
+UsingExistingConnectionsCheaperThanEstablishingNewConnections(int newConnectionCount,
+															  WorkerPool *workerPool)
+{
+	int activeConnectionCount = workerPool->activeConnectionCount;
+
+	if (workerPool->totalExecutedTasks < 1 || activeConnectionCount < 1)
+	{
+		/*
+		 * The pool has not finished any connection establishment or
+		 * task yet. So, we refrain from optimizing the execution.
+		 */
+		return false;
+	}
+
+	double avgTaskExecutionTime = AvgTaskExecutionTimeApproximation(workerPool);
+	double avgConnectionEstablishmentTime = AvgConnectionEstablishmentTime(workerPool);
+
+	/* we assume that we are halfway through the execution */
+	double remainingTimeForActiveTaskExecutionsToFinish = avgTaskExecutionTime / 2;
+
+	/*
+	 * We use "newConnectionCount" as if it is the task count as
+	 * we are only interested in this iteration of CalculateNewConnectionCount().
+	 */
+	int taskCount = newConnectionCount;
+	double totalTimeToExecuteNewTasks = avgTaskExecutionTime * taskCount;
+
+	double estimatedExecutionTimeForNewTasks =
+		floor(totalTimeToExecuteNewTasks / activeConnectionCount);
+
+	/*
+	 * First finish the already running tasks, and then use the connections
+	 * to execute the new tasks.
+	 */
+	double costOfExecutingTheTasksOverExistingConnections =
+		remainingTimeForActiveTaskExecutionsToFinish +
+		estimatedExecutionTimeForNewTasks;
+
+	/*
+	 * For every task, the executor is supposed to establish one
+	 * connection and then execute the task over the connection.
+	 */
+	double costOfExecutingTheTasksOverNewConnection =
+		(avgTaskExecutionTime + avgConnectionEstablishmentTime);
+
+	return (costOfExecutingTheTasksOverExistingConnections <=
+			costOfExecutingTheTasksOverNewConnection);
+}
+
+
+/*
+ * AvgTaskExecutionTimeApproximation returns the approximation of the average task
+ * execution times on the workerPool.
+ */
+static double
+AvgTaskExecutionTimeApproximation(WorkerPool *workerPool)
+{
+	double totalTaskExecutionTime = workerPool->totalTaskExecutionTime;
+	int taskCount = workerPool->totalExecutedTasks;
+
+	instr_time now;
+	INSTR_TIME_SET_ZERO(now);
+
+	WorkerSession *session = NULL;
+	foreach_ptr(session, workerPool->sessionList)
+	{
+		/* only set now once */
+		if (INSTR_TIME_IS_ZERO(now))
+		{
```

Comment:
why not just set it above? sessionList is almost never going to be empty

Reply:
makes sense